### PR TITLE
Extract updateSession function to reduce repetition

### PR DESCRIPTION
**Pull request by wolfogre**
A simple refactor to reduce duplicate codes.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: routers/web/auth/auth.go [hunk 1]
```diff
@@ -814,3 +779,25 @@ func ActivateEmail(ctx *context.Context) {
 	// Should users be logged in automatically here? (consider 2FA requirements, etc.)
 	ctx.Redirect(setting.AppSubURL + "/user/settings/account")
 }
+
+func updateSession(ctx *context.Context, deletes []interface{}, updates map[interface{}]interface{}) error {
```
Comment (KN4CK3R) [inline routers/web/auth/auth.go]:
```suggestion
type sessionData = map[string]interface{}

func updateSession(ctx *context.Context, keysToDelete []string, updates sessionData) error {
```
We only use strings as key, so for now we could keep this simple. The `sessionData` type is to reduce the noise when calling the method. What do you think?

```go
updateSession(ctx, []string{
	// Delete the openid, 2fa and linkaccount data
	"openid_verified_uri",
	"openid_signin_remember",
	"openid_determined_email",
	"openid_determined_username",
	"twofaUid",
	"twofaRemember",
	"linkAccount",
}, sessionData{
	"uid":   u.ID,
	"uname": u.Name,
})
```

Comment (wolfogre) [reply routers/web/auth/auth.go]: It looks better, but I'm not sure if we need to define a new type, maybe `map[string]interface{}` is good enough?

Comment (KN4CK3R) [reply routers/web/auth/auth.go]: Sure, it's just less to type.

**Review by KN4CK3R — APPROVED** [review]

File: routers/web/auth/auth.go [hunk 2]
```diff
@@ -821,3 +786,25 @@ func ActivateEmail(ctx *context.Context) {
...
+	sess := ctx.Session
+	sessID := sess.ID()
+	for _, k := range deletes {
+		if err := sess.Delete(k); err != nil {
+			return fmt.Errorf("delete %v in session[%s]: %v", k, sessID, err)
+		}
+	}
+	for k, v := range updates {
+		if err := sess.Set(k, v); err != nil {
+			return fmt.Errorf("set %v in session[%s]: %v", k, sessID, err)
```
Comment (delvh) [inline routers/web/auth/auth.go]:
```suggestion
			return fmt.Errorf("delete %v in session[%s]: %w", k, sessID, err)
		}
	}
	for k, v := range updates {
		if err := sess.Set(k, v); err != nil {
			return fmt.Errorf("set %v in session[%s]: %w", k, sessID, err)
```

File: routers/web/auth/auth.go [hunk 3]
```diff
@@ -82,21 +82,14 @@ func AutoSignIn(ctx *context.Context) (bool, error) {
 
 	isSucceed = true
 
-	if _, err := session.RegenerateSession(ctx.Resp, ctx.Req); err != nil {
+	if err := updateSession(ctx, nil, map[string]interface{}{
+		// Set session IDs
+		"uid":   u.ID,
+		"uname": u.Name,
+	}); err != nil {
 		return false, fmt.Errorf("unable to RegenerateSession: Error: %w", err)
```
Comment (delvh) [inline routers/web/auth/auth.go]:
```suggestion
		return false, fmt.Errorf("unable to updateSession: %w", err)
```

File: routers/web/auth/auth.go [hunk 4]
```diff
@@ -252,32 +245,17 @@ func SignInPost(ctx *context.Context) {
...
-			ctx.ServerError("UserSignIn: Unable to set WebAuthn Enrolled in session", err)
-			return
-		}
+		updates["totpEnrolled"] = u.ID
 	}
-
-	if err := ctx.Session.Release(); err != nil {
-		ctx.ServerError("UserSignIn: Unable to save session", err)
+	if err := updateSession(ctx, nil, updates); err != nil {
+		ctx.ServerError("UserSignIn: Unable to set regenerate session", err)
```
Comment (delvh) [inline routers/web/auth/auth.go]:
```suggestion
		ctx.ServerError("UserSignIn: Unable to update session", err)
```

File: routers/web/auth/oauth.go [hunk 5]
```diff
@@ -1075,21 +1069,14 @@ func handleOAuth2SignIn(ctx *context.Context, source *auth.Source, u *user_model
 	// If this user is enrolled in 2FA and this source doesn't override it,
 	// we can't sign the user in just yet. Instead, redirect them to the 2FA authentication page.
 	if !needs2FA {
-		if _, err := session.RegenerateSession(ctx.Resp, ctx.Req); err != nil {
+		if err := updateSession(ctx, nil, map[string]interface{}{
+			"uid":   u.ID,
+			"uname": u.Name,
+		}); err != nil {
 			ctx.ServerError("RegenerateSession", err)
```
Comment (delvh) [inline routers/web/auth/oauth.go]:
```suggestion
			ctx.ServerError("updateSession", err)
```

File: routers/web/auth/oauth.go [hunk 6]
```diff
@@ -1138,22 +1125,15 @@ func handleOAuth2SignIn(ctx *context.Context, source *auth.Source, u *user_model
 		}
 	}
 
-	if _, err := session.RegenerateSession(ctx.Resp, ctx.Req); err != nil {
+	if err := updateSession(ctx, nil, map[string]interface{}{
+		// User needs to use 2FA, save data and redirect to 2FA page.
+		"twofaUid":      u.ID,
+		"twofaRemember": false,
+	}); err != nil {
 		ctx.ServerError("RegenerateSession", err)
```
Comment (delvh) [inline routers/web/auth/oauth.go]:
```suggestion
		ctx.ServerError("updateSession", err)
```

File: routers/web/auth/openid.go [hunk 7]
```diff
@@ -232,27 +231,16 @@ func signInOpenIDVerify(ctx *context.Context) {
...
-		log.Error("signInOpenIDVerify: Could not set openid_determined_username in session: %v", err)
-	}
-	if err := ctx.Session.Release(); err != nil {
-		log.Error("signInOpenIDVerify: Unable to save changes to the session: %v", err)
+	if err := updateSession(ctx, nil, map[string]interface{}{
+		"openid_verified_uri":        id,
+		"openid_determined_email":    email,
+		"openid_determined_username": nickname,
+	}); err != nil {
+		ctx.ServerError("RegenerateSession", err)
```
Comment (delvh) [inline routers/web/auth/openid.go]:
```suggestion
		ctx.ServerError("updateSession", err)
```

File: routers/web/auth/oauth.go [hunk 8]
```diff
@@ -1027,17 +1026,12 @@ func setUserGroupClaims(loginSource *auth.Source, u *user_model.User, gothUser *
 }
 
 func showLinkingLogin(ctx *context.Context, gothUser goth.User) {
-	if _, err := session.RegenerateSession(ctx.Resp, ctx.Req); err != nil {
+	if err := updateSession(ctx, nil, map[string]interface{}{
+		"linkAccountGothUser": gothUser,
+	}); err != nil {
 		ctx.ServerError("RegenerateSession", err)
```
Comment (delvh) [inline routers/web/auth/oauth.go]:
```suggestion
		ctx.ServerError("updateSession", err)
```

**Comment by delvh** [comment]
> **Note**

I didn't touch the meaningfulness of these error messages, I simply restored consistency a bit.

**Review by delvh — APPROVED** [review]

**Pull request merged**